hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/recordcache/RecordCacheImpl.java

```diff
@@ -257,33 +266,39 @@ private void removeExpiredTransactions(
             @NonNull final Timestamp consensusTimestamp) {
         // Compute the earliest valid start timestamp that is still within the max transaction duration window.
         final var config = configProvider.getConfiguration().getConfigData(HederaConfig.class);
-        final var earliestValidState = minus(consensusTimestamp, config.transactionMaxValidDuration());
-
-        // Loop in order and expunge every entry where the timestamp is before the current time. Also remove from the
-        // in memory data structures.
-        final var itr = queue.iterator();
-        while (itr.hasNext()) {
-            final var entry = itr.next();
-            final var rec = entry.transactionRecordOrThrow();
-            final var txId = rec.transactionIDOrThrow();
-            // If the timestamp is before the current time, then it has expired
-            if (isBefore(txId.transactionValidStartOrThrow(), earliestValidState)) {
-                // Remove from the histories
-                itr.remove();
-                // Remove from the payer to transaction index
-                final var payerAccountId = txId.accountIDOrThrow(); // NOTE: Not accurate if the payer was the node
-                final var transactionIDs =
-                        payerToTransactionIndex.computeIfAbsent(payerAccountId, ignored -> new HashSet<>());
-                transactionIDs.remove(txId);
-                if (transactionIDs.isEmpty()) {
-                    payerToTransactionIndex.remove(payerAccountId);
+        final var earliestValidStart = minus(consensusTimestamp, config.transactionMaxValidDuration());
+        // Loop in order and expunge every entry where the start time is before the earliest valid start.
+        // Also remove from the in-memory data structures.
+        do {
+            final var entry = queue.peek();
+            if (entry != null) {
+                final var rec = entry.transactionRecordOrThrow();
+                final var txId = rec.transactionIDOrThrow();
+                // If the valid start time is before the earliest valid start, then it has expired
+                if (isBefore(txId.transactionValidStartOrThrow(), earliestValidStart)) {
+                    // Remove from the histories.  Note that all transactions are added to this map
+                    // keyed to the "user transaction" ID, so removing the entry here removes both
+                    // "parent" and "child" transaction records associated with that ID.
+                    histories.remove(txId);
+                    // remove from queue as well.  The queue only permits removing the current "HEAD",
+                    // but that should always be correct here.
+                    queue.removeIf(TruePredicate.INSTANCE);
```

Comment:
Will this empty everything out of the queue?

Was there something functionally incorrect in the previous code, or was it just wasteful creating the iterator? What is the purpose to this change?

Reply:
There is no iterator for FCQueue, it can only be removed from the head, and nowhere else.

Reply:
I ran into this yesterday, that the `remove()` operation wasn't implemented. I'm really curious if purging transactions has been working at all.

Reply:
Transactions were never removed from anything before this PR. The queue was always empty, so we never entered the remove loop, because we weren't committing the state.

The two bugs that this PR fixes are, paraphrased, "queue in state is always empty" and "histories leaks memory because nothing is ever removed from that map".